clients/ssh/git_upload_pack.go

```diff
@@ -0,0 +1,194 @@
+package ssh
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+
+	"github.com/sourcegraph/go-vcsurl"
+
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+
+	"gopkg.in/src-d/go-git.v2/clients/common"
+	"gopkg.in/src-d/go-git.v2/formats/pktline"
+)
+
+type GitUploadPackService struct {
+	connected bool
+	vcs       *vcsurl.RepoInfo
+	client    *ssh.Client
+}
+
+func NewGitUploadPackService() *GitUploadPackService {
+	return &GitUploadPackService{}
+}
+
+func (s *GitUploadPackService) Connect(ep common.Endpoint) (err error) {
+	if s.connected {
+		return fmt.Errorf("already connected")
+	}
+
+	s.vcs, err = vcsurl.Parse(string(ep))
+	if err != nil {
+		return fmt.Errorf("cannot parse vcs endpoint: %v", err)
+	}
+
+	url, err := vcsToUrl(s.vcs)
+	if err != nil {
+		return
+	}
+
+	s.client, err = connect(url.Host, url.User.Username())
+	if err != nil {
+		return fmt.Errorf("cannot connect: %v")
+	}
+	s.connected = true
+	return
+}
+
+func vcsToUrl(vcs *vcsurl.RepoInfo) (u *url.URL, err error) {
+	if vcs.VCS != vcsurl.Git {
+		return nil, fmt.Errorf("only git repos are supported, found %s", vcs.VCS)
+	}
+	if vcs.RepoHost != vcsurl.GitHub {
+		return nil, fmt.Errorf("only github.com host is supported, found %s", vcs.RepoHost)
+	}
+	s := "ssh://git@" + string(vcs.RepoHost) + ":22/" + vcs.FullName
+	u, err = url.Parse(s)
+	return
+}
+
+func connect(host, user string) (*ssh.Client, error) {
+
+	// try ssh-agent first
+	// then password auth
+	conn, err := net.Dial("unix", os.Getenv("SSH_AUTH_SOCK"))
+	if err != nil {
+		return connectWithPasswd(host, user)
+	}
+	defer conn.Close()
+
+	agent := agent.NewClient(conn)
+	sshConfig := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{ssh.PublicKeysCallback(agent.Signers)},
+	}
+
+	client, err := ssh.Dial("tcp", host, sshConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial: %v", err)
+	}
+
+	return client, nil
+}
+
+// TODO: this must be done securely
+func connectWithPasswd(host, user string) (*ssh.Client, error) {
+	var pass string = "your password"
+
+	sshConfig := &ssh.ClientConfig{
+		User: user,
+		Auth: []ssh.AuthMethod{ssh.Password(pass)},
+	}
+
+	client, err := ssh.Dial("tcp", host, sshConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dia: %v", err)
+	}
+
+	return client, nil
+}
+
+func (s *GitUploadPackService) Info() (i *common.GitUploadPackInfo, err error) {
+	if !s.connected {
+		return nil, fmt.Errorf("not connected")
```

Comment:
Just in this case, because it's the same as `errors.New("not connected")`, you can make the error a public variable, like `ErrNotConnected` so library users can check against it. All other errors need a common error type.